content/docs/user-guide/dvc-files.md

```diff
@@ -240,14 +248,245 @@ stages:
+[parameters files](/doc/command-reference/params), or internal `vars`, as
+detailed below.
+
+Let's say we have `params.yaml` (default params file) with the following
+contents:
+
+```yaml
+models:
+  us:
+    threshold: 10
+    filename: 'model-us.hdf5'
+```
+
+Those values can be used anywhere in `dvc.yaml` with the `${}` _substitution
+expression_:
+
+<!-- prettier-ignore-start -->
+```yaml
+stages:
+  build-us:
+    cmd: >-
+      python train.py
+      --thresh ${models.us.threshold}
+      --out ${models.us.filename}
+    outs:
+      - ${models.us.filename}:
+        cache: true
+```
+<!-- prettier-ignore-end -->
+
+DVC will track simple param values (numbers, strings, etc.) used in `${}` (they
+will be listed by `dvc params diff`).
+
+Alternatively, values for substitution can be listed as top-level `vars` like
+this:
+
+```yaml
+vars:
+  - models:
+      us:
+        threshold: 10
+  - desc: 'Reusable description'
+
+stages:
+  build-us:
+    desc: ${desc}
+    cmd: python train.py --thresh ${models.us.threshold}
+```
+
+> DVC merges external params and `vars` as long as there are no leaf node
+> collisions (so the two examples above can't be used simultaneously). For
+> example, `{"grp": {"a": 1}}` can be merged with `{"grp": {"b": 2}}`, but not
+> with `{"grp": {"a": 7}}`.
+
+⚠️ Note that doesn't track values from `vars` like parameters (ignored by
+`dvc params diff`).
+
+To load additional params files, list them in the top `vars`, in the desired
+order, e.g.:
+
+> Their paths will be evaluated based on `wdir`, if one given.
+
+```yaml
+vars:
+  - params.json
+  - myvar: 'value'
+  - config/myapp.yaml
+```
+
+(ℹ️) Note that the default `params.yaml` file is always loaded first, if
+present.
+
+It's also possible to specify what to include from additional params files, with
+a `:` colon:
+
+```yaml
+vars:
+  - params.json:clean,feats
+
+stages:
+  featurize:
+    cmd: ${feats.exec}
+    deps:
+      - ${clean.filename}
+    outs:
+      - ${feats.dirname}
+```
+
+Stage-specific values are also supported, with inner `vars`. You may also load
+additional params files locally. For example:
+
+```yaml
+stages:
+  build-us:
+    vars:
+      - params.json:build
+      - model:
+        filename: 'model-us.hdf5'
+    cmd: python train.py ${build.epochs} --out ${model.filename}
+    outs:
+      - ${model.filename}
+```
+
+⚠️ Important: Limitations of local `vars`:
+
+- `wdir` cannot use values from local `vars`, as DVC uses the working directory
+  first (to load any values from params files listed in `vars`).
+- `foreach` is also incompatible with local `vars` at the moment.
+
+The substitution expression supports these forms:
+
+```yaml
+${param} # Simple
+${param.key} # Nested values through . (period)
+${param.list[0]} # List elements via index in [] (square brackets)
+```
+
+> To use the expression literally in `dvc.yaml`, escape it with a backslash,
+> e.g. `\${...`.
+
+### Defining multiple stages at once
```

Comment:
how about `foreach`/`do` - using loops to generate stages

Reply:
I'm not sure we want to introduce the concept of "loops" (even when its implied by the `foreach` analogy). Per https://github.com/iterative/dvc/issues/5181#issuecomment-752326990 the items given to `foreach` are not ordered so it doesn't seem like an actual loop to me.

Reply:
\* not considered ordered

Reply:
(This title is **Generating multiple stages** now BTW.)